File: backend/app/services/skill_gap_engine.py
from typing import List, Dict, Any, Optional

def calculate_role_readiness(
    student_skills: Optional[List[dict]] = None,
    target_role_skills: Optional[List[dict]] = None,
    role_skills: Optional[List[dict]] = None,
    **kwargs
) -> dict:
    """
    Computes role readiness score, radar points, and classifies skills:
    - Strength: Proficiency >= Required benchmark
    - Moderate Gap: 50% <= Proficiency < Required benchmark
    - Critical Gap: Proficiency < 50% of benchmark (or missing mandatory skill)
    """
    role_skill_list = target_role_skills or role_skills or []
    student_skill_list = student_skills or kwargs.get("student_skill_list") or []

    if not role_skill_list:
        return {
            "readiness_score": 0.0,
            "readiness_percentage": 0.0,
            "general_skill_score": 0.0,
            "radar_data": [],
            "critical_gaps": [],
            "moderate_gaps": [],
            "strengths": []
        }

    skills_map = {s.get("skill_id"): s for s in student_skill_list if s.get("skill_id")}
    
    radar_data = []
    critical_gaps = []
    moderate_gaps = []
    strengths = []

    total_weighted_score = 0.0
    total_max_weight = 0.0
    sum_prof = 0.0

    for rs in role_skill_list:
        s_id = rs.get("skill_id")
        req_prof = float(rs.get("required_proficiency", 70.0))
        weight = float(rs.get("weight", 1.0))
        is_mandatory = bool(rs.get("is_mandatory", True))
        skill_name = rs.get("skill_name", "Skill")
        category_name = rs.get("category_name", rs.get("category", "Technical"))

        st_skill = skills_map.get(s_id)
        current_prof = float(st_skill.get("proficiency_score", 0.0)) if st_skill else 0.0
        verification = st_skill.get("verification_level", "NONE") if st_skill else "NONE"
        sum_prof += current_prof

        # Gap calculation
        gap = max(0.0, req_prof - current_prof)
        ratio = min(1.0, current_prof / max(req_prof, 1.0))
        
        # Mandatory skills have higher penalty
        effective_weight = weight * (1.5 if is_mandatory else 1.0)
        total_weighted_score += ratio * effective_weight
        total_max_weight += effective_weight

        item = {
            "skill_id": s_id,
            "name": skill_name,
            "category": category_name,
            "proficiency_score": current_prof,
            "verification_level": verification,
            "required_proficiency": req_prof,
            "gap": round(gap, 1),
        }

        radar_data.append({
            "skill": skill_name[:18],
            "student": current_prof,
            "required": req_prof
        })

        if current_prof >= req_prof:
            item["status"] = "STRENGTH"
            strengths.append(item)
        elif current_prof >= (req_prof * 0.5):
            item["status"] = "MODERATE_GAP"
            moderate_gaps.append(item)
        else:
            item["status"] = "CRITICAL_GAP"
            critical_gaps.append(item)

    readiness = round((total_weighted_score / total_max_weight) * 100.0, 1) if total_max_weight > 0 else 0.0
    general_skill = round(sum_prof / max(len(role_skill_list), 1), 1)

    return {
        "readiness_score": readiness,
        "readiness_percentage": readiness,
        "general_skill_score": general_skill,
        "radar_data": radar_data,
        "critical_gaps": critical_gaps,
        "moderate_gaps": moderate_gaps,
        "strengths": strengths
    }

File: frontend/src/lib/services/skillGapEngine.ts
// Skill Gap Engine
// Computes role-specific gaps and readiness benchmarks

export type GapCategory = "Strong Match" | "Minor Gap" | "Moderate Gap" | "Major Gap";

export interface SkillGapItem {
  skillId: string;
  skillName: string;
  categoryName?: string;
  studentScore: number;
  requiredScore: number;
  gap: number;
  classification: GapCategory;
  isMandatory: boolean;
  weight: number;
}

export interface RoleReadinessResult {
  roleId: string;
  roleTitle: string;
  readinessPercentage: number;
  strengths: SkillGapItem[];
  minorGaps: SkillGapItem[];
  moderateGaps: SkillGapItem[];
  majorGaps: SkillGapItem[];
  allGaps: SkillGapItem[];
}

export function classifyGap(gap: number): GapCategory {
  if (gap <= 5) return "Strong Match";
  if (gap <= 15) return "Minor Gap";
  if (gap <= 30) return "Moderate Gap";
  return "Major Gap";
}

/**
 * Calculates skill gap: Industry Requirement - Student Verified Skill
 * If student has higher than required, gap is 0 (Strong Match)
 */
export function calculateSkillGap(requiredScore: number, studentScore: number) {
  const rawDiff = requiredScore - studentScore;
  const gap = Math.max(0, Math.round(rawDiff * 10) / 10);
  const classification = classifyGap(gap);
  return { gap, classification };
}

/**
 * Calculates student readiness for a specific Career Role based on database-mapped skills
 */
export function calculateRoleReadiness(
  role: {
    id: string;
    title: string;
    skills: Array<{
      skillId: string;
      skill?: { id?: string; name?: string; category?: { name?: string } };
      skillName?: string;
      requiredProficiency: number;
      isMandatory: boolean;
      weight: number;
    }>;
  },
  studentSkills: Array<{
    skillId: string;
    proficiencyScore: number;
  }>
): RoleReadinessResult {
  const studentSkillMap = new Map<string, number>();
  for (const s of studentSkills) {
    studentSkillMap.set(s.skillId, s.proficiencyScore);
  }

  const items: SkillGapItem[] = [];
  let totalWeightedScore = 0;
  let totalWeight = 0;

  for (const req of (role.skills ?? [])) {
    if (!req) continue;
    const studentScore = studentSkillMap.get(req.skillId) || 0;
    const requiredScore = req.requiredProficiency || 70;
    const { gap, classification } = calculateSkillGap(requiredScore, studentScore);

    // Scoring: Ratio of student score to required score (capped at 1.0)
    // If mandatory and student score is 0, penalty applies
    const ratio = requiredScore > 0 ? Math.min(1.0, studentScore / requiredScore) : 1.0;
    const weight = req.weight || 1.0;

    totalWeightedScore += ratio * 100 * weight;
    totalWeight += weight;

    items.push({
      skillId: req.skillId,
      skillName: req.skill?.name ?? req.skillName ?? "Skill",
      categoryName: req.skill?.category?.name,
      studentScore,
      requiredScore,
      gap,
      classification,
      isMandatory: req.isMandatory ?? false,
      weight,
    });
  }

  const readinessPercentage =
    totalWeight > 0 ? Math.round(totalWeightedScore / totalWeight) : 0;

  const strengths = items.filter((i) => i.classification === "Strong Match");
  const minorGaps = items.filter((i) => i.classification === "Minor Gap");
  const moderateGaps = items.filter((i) => i.classification === "Moderate Gap");
  const majorGaps = items.filter((i) => i.classification === "Major Gap");

  return {
    roleId: role.id,
    roleTitle: role.title,
    readinessPercentage,
    strengths,
    minorGaps,
    moderateGaps,
    majorGaps,
    allGaps: items,
  };
}
